### PR TITLE
Support stack unwind on Android

### PR DESCRIPTION
**Pull request by huangqinjin**

### CODE REVIEW

**Comment by sergiud** [comment]
Thanks for the contribution. However, many of the changes are more than questionable. Therefore, please provide a description of the purpose of this PR.

File: CMakeLists.txt [hunk 1]
```diff
@@ -89,7 +89,8 @@ find_package (Unwind)
 
 if (Unwind_FOUND)
   set (HAVE_LIB_UNWIND 1)
-  set (HAVE_UNWIND_H 1)
+else (Unwind_FOUND)
+  check_include_file_cxx (unwind.h HAVE_UNWIND_H)
```
Comment (sergiud) [inline CMakeLists.txt:93]: Why do you need this? `FindUnwind.cmake` already accomplishes this.

File: bazel/glog.bzl [hunk 2]
```diff
@@ -128,7 +128,7 @@ def glog_library(namespace = "google", with_gflags = 1, **kwargs):
             "src/stacktrace_powerpc-inl.h",
             "src/stacktrace_windows-inl.h",
             "src/stacktrace_x86-inl.h",
-            "src/stacktrace_x86_64-inl.h",
```
Comment (sergiud) [inline bazel/glog.bzl:131]: Why did you remove the header from the list?

File: cmake/FindUnwind.cmake [hunk 3]
```diff
@@ -6,8 +6,8 @@
 
 include (FindPackageHandleStandardArgs)
 
-find_path (Unwind_INCLUDE_DIR NAMES unwind.h libunwind.h DOC "unwind include directory")
-find_library (Unwind_LIBRARY NAMES unwind DOC "unwind library")
+find_path (Unwind_INCLUDE_DIR NAMES libunwind.h DOC "libunwind include directory")
```
Comment (sergiud) [inline cmake/FindUnwind.cmake:9]: This is a breaking change that cannot be accepted.

File: src/logging.cc [hunk 4]
```diff
@@ -779,7 +779,8 @@ static void WriteToStderr(const char* message, size_t len) {
 }
 
 inline void LogDestination::MaybeLogToStderr(LogSeverity severity,
-					     const char* message, size_t message_len, size_t /*prefix_len*/) {
+					     const char* message, size_t message_len, size_t prefix_len) {
```
Comment (sergiud) [inline src/logging.cc:782]: The parameter name is uncommented but still unused. What's the purpose of this change then?

File: src/logging.cc [hunk 5]
```diff
@@ -1868,6 +1846,11 @@ void LogMessage::SendToLog() EXCLUSIVE_LOCKS_REQUIRED(log_mutex) {
     if (write(STDERR_FILENO, message, strlen(message)) < 0) {
       // Ignore errors.
     }
+#if defined(__ANDROID__)
+    __android_log_write(ANDROID_LOG_FATAL,
```
Comment (sergiud) [inline src/logging.cc:1850]: It is not clear why messages written to logcat are always of fatal severity.

File: src/utilities.cc [hunk 6]
```diff
@@ -92,6 +95,11 @@ static void DebugWriteToStderr(const char* data, void *) {
   if (write(STDERR_FILENO, data, strlen(data)) < 0) {
     // Ignore errors.
   }
+#if defined(__ANDROID__)
+  __android_log_write(ANDROID_LOG_FATAL,
```
Comment (sergiud) [inline src/utilities.cc:99]: As before: it is not clear why messages written to logcat are always of fatal severity.

**Review by sergiud — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (huangqinjin) [reply CMakeLists.txt:93]: `FindUnwind.cmake` is for libunwind library. It has `libunwind.h` and `libunwind.so`. And its API is `unw_*` and is low level. `unwind.h` contains high level API `_Unwind_*` (based on `unw_*` in case of [LLVM](https://github.com/llvm/llvm-project/tree/main/libunwind/include)).

File: bazel/glog.bzl (hunk 2 above)
Comment (huangqinjin) [reply bazel/glog.bzl:131]: I renamed it to `stacktrace_unwind-inl.h` as it actually uses `_Unwind_*` API rather than hand-coded like `stacktrace_x86-inl.h`.

File: CMakeLists.txt (hunk 1 above)
Comment (sergiud) [reply CMakeLists.txt:93]: Please check #629 for a related discussion before making further adjustments.

File: src/logging.cc (hunk 4 above)
Comment (huangqinjin) [reply src/logging.cc:782]: For Android. https://github.com/google/glog/blob/9733727be4aed835e21a8f73c403f546b71663fe/src/logging.cc#L798

File: src/utilities.cc (hunk 6 above)
Comment (huangqinjin) [reply src/utilities.cc:99]: Because here is fatal error and is going to print stack trace.

File: src/logging.cc (hunk 5 above)
Comment (huangqinjin) [reply src/logging.cc:1850]: Because here is fatal error and is going to print stack trace.

File: src/logging.cc (hunk 4 above)
Comment (sergiud) [reply src/logging.cc:782]: I see.  Could you please then move the `void` expression to the `#if` block that does not use the parameter?

File: src/logging.cc (hunk 5 above)
Comment (sergiud) [reply src/logging.cc:1850]: Could please add this as comment above then?

File: CMakeLists.txt (hunk 1 above)
Comment (huangqinjin) [reply CMakeLists.txt:93]: `HAVE_LIB_UNWIND` includes `libunwind.h` and `HAVE_UNWIND_H` includes `unwind.h`. They are irrelevant in current code:

https://github.com/google/glog/blob/9733727be4aed835e21a8f73c403f546b71663fe/src/utilities.h#L108-L114

File: cmake/FindUnwind.cmake (hunk 3 above)
Comment (huangqinjin) [reply cmake/FindUnwind.cmake:9]: Here we search for `libunwind.h` and `libunwind.so` and will define `HAVE_LIB_UNWIND` if found and use `unw_*` API in code. `unwind.h` does not declare `unw_*` but `_Unwind_*`.

File: CMakeLists.txt (hunk 1 above)
Comment (huangqinjin) [reply CMakeLists.txt:93]: Though Android NDK uses LLVM, only `unwind.h` is shipped.

File: cmake/FindUnwind.cmake (hunk 3 above)
Comment (sergiud) [reply cmake/FindUnwind.cmake:9]: This may be true. However, `FindUnwind.cmake` is intended to locate `unwind.h` not `libunwind.h`. The former is included by `stacktrace_x86_64-inl.h`:
https://github.com/google/glog/blob/9733727be4aed835e21a8f73c403f546b71663fe/src/stacktrace_x86_64-inl.h#L35

This is why I argue this change is not acceptable.

Comment (huangqinjin) [reply cmake/FindUnwind.cmake:9]: > FindUnwind.cmake is intended to locate unwind.h not libunwind.h

It is not the current logic.

https://github.com/google/glog/blob/9733727be4aed835e21a8f73c403f546b71663fe/CMakeLists.txt#L88-L93


And  `HAVE_LIB_UNWIND` takes precedence over `HAVE_UNWIND_H`. So `unwind.h` is not used if `Unwind_FOUND`.

https://github.com/google/glog/blob/9733727be4aed835e21a8f73c403f546b71663fe/src/utilities.h#L108-L114

Comment (sergiud) [reply cmake/FindUnwind.cmake:9]: Okay, but how are these changes (and renaming of headers) are related to Android support?

Comment (huangqinjin) [reply cmake/FindUnwind.cmake:9]: only `unwind.h` is shipped with Android NDK.

Comment (huangqinjin) [reply cmake/FindUnwind.cmake:9]: I renamed `stacktrace_x86_64-inl.h` to `stacktrace_unwind-inl.h` as it actually uses `_Unwind_*` API rather than hand-coded like `stacktrace_x86-inl.h`.

Comment (huangqinjin) [reply cmake/FindUnwind.cmake:9]: Changes to `FindUnwind.cmake` actually are not related to Android support. Just correct it.

Comment (sergiud) [reply cmake/FindUnwind.cmake:9]: Got it. I prefer moving `FindUnwind.cmake` changes to a separate PR.

Comment (huangqinjin) [reply cmake/FindUnwind.cmake:9]: ok. what about the renaming change?

Comment (sergiud) [reply cmake/FindUnwind.cmake]: I would revert that too for now. But I agree that we should probably rename the header to make its purpose clearer. Maybe in the same PR as changes to `FindUnwind.cmake`.

**Comment by sergiud** [comment]
Is there a chance you could setup a Github action that uses (any) Android NDK? Without a CI pipeline, it is very likely that the Android support will break again.

**Comment by huangqinjin** [comment]
OK, I will make a separate PR for Android CI.

**Comment by sergiud** [comment]
Thank you. A separate PR is not needed though. You can add the Github action as part of this PR. This will also ensure your changes actually work.

**Comment by huangqinjin** [comment]
CI added. I disabled tests since running tests requires emulators.

File: .github/workflows/android-build.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        cmake -S . -B ${{runner.workspace}}/build_CXX${{matrix.std}}-${{matrix.abi}} \
+              -G "Ninja Multi-Config" \
+              -DCMAKE_CXX_EXTENSIONS=OFF \
+              -DCMAKE_CXX_STANDARD=${{matrix.std}} \
+              -DCMAKE_CXX_STANDARD_REQUIRED=ON \
+              -DCMAKE_TOOLCHAIN_FILE=${ANDROID_NDK_HOME}/build/cmake/android.toolchain.cmake \
+              -DANDROID_STL=c++_shared \
+              -DANDROID_NATIVE_API_LEVEL=24 \
+              -DANDROID_ABI=${{matrix.abi}} \
+              -DBUILD_TESTING=OFF
```
Comment (sergiud) [inline .github/workflows/android-build.yml:36]: I would still build the tests to ensure at least linking works.

File: .github/workflows/android-build.yml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+jobs:
+  build:
+    name: NDK-C++${{matrix.std}}-${{matrix.abi}}-${{matrix.build_type}}
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+          std: [98, 11, 14, 17, 20]
+          abi: [arm64-v8a, armeabi-v7a, x86_64, x86]
+          build_type: [Debug, Release, RelWithDebInfo]
```
Comment (sergiud) [inline .github/workflows/android-build.yml:14]: Is `RelWithDebInfo` actually necessary? I'd say `Debug` and maybe `Release` are sufficient.

File: CMakeLists.txt (hunk 1 above)
Comment (sergiud) [inline CMakeLists.txt:93]: I still do not understand the change. `FindUnwind.cmake` can already tell if `unwind.h` is present.

**Comment by sergiud** [comment]
Thanks!

**Review by sergiud — APPROVED** [review]
